# Hypothesis testing problems

## Exercise 1

You are a nutritionist researching two different types of diets to see if there is a significant difference in weight loss after one month. You choose two random groups of people; one group follows the first diet, and the other follows the second. At the end of the month, you record the weight loss (in kg) for each person:

| Diet 1 | Diet 2 |
|:-------|:-------|
| 2.0 | 3.0 |
| 2.5 | 3.2 |
| 3.0 | 3.1 |
| 2.8 | 2.9 |
| 2.3 | 2.8 |
| 2.7 | 3.0 |
| 2.5 | 3.2 |

With these data, it seeks to answer the following question: Is there a significant difference in average weight loss between people who followed the first diet and those who followed the second diet?

To draw conclusions, follow the points below:

- State the hypothesis: null and alternative hypothesis.
- Perform the test to test the hypothesis. You can use a Student's t-test.
- Analyze the conclusions.

In [48]:
# TODO
import pandas as pd
import numpy as np

diet1 = np.array([2.0,2.5,3.0,2.8,2.3,2.7,2.5])
diet2 = np.array([3.0,3.2,3.1,2.9,2.8,3.0,3.0])
data = pd.DataFrame([diet1,diet2])#, columns = ["Diet 1", "Diet 2"])
data = data.T
data.columns = ("Diet 1", "Diet 2")
data

Diet 1  Diet 2
0     2.0     3.0
1     2.5     3.2
2     3.0     3.1
3     2.8     2.9
4     2.3     2.8
5     2.7     3.0
6     2.5     3.0

Hypothesis (null):

-Results between Diet 1 and Diet 2 will not be significantly different

Hypothesis 1:

-Diet 1 and Diet 2 will have significantly different results.

In [49]:
from scipy.stats import ttest_ind as t

results = t(a=diet1, b=diet2)
print(results)
print(f"Diet 1's mean value is {round(diet1.mean(),2)}")
print(f"Diet 2's mean value is {round(diet2.mean(),2)}")


TtestResult(statistic=np.float64(-3.4047691302331584), pvalue=np.float64(0.005223479648626318), df=np.float64(12.0))
Diet 1's mean value is 2.54
Diet 2's mean value is 3.0


In [62]:
"""
To analyze the differences between the results of both diets a Two Sample Student's T Test was performed.
"""

print("Conclusions:")
print(f"Results from the Two Sample Student's T Test show that the difference in outcomes between the two diets is very significant, with a p value lesser than .05 ({round(results[1],3)})). Therefore, confirming Hyphothesis 1. \n")
print(f"Analyzing the mean scores of each sample, we can confirm that Diet 2's mean value ({round(diet2.mean(),2)}) is greater than Diet 1's mean value ({round(diet1.mean(),2)}). \nThis way we can conclude that Diet 2 was significantly more effective than Diet 1.")

Conclusions:
Results from the Two Sample Student's T Test show that the difference in outcomes between the two diets is very significant, with a p value lesser than .05 (0.005)). Therefore, confirming Hyphothesis 1. 

Analyzing the mean scores of each sample, we can confirm that Diet 2's mean value (3.0) is greater than Diet 1's mean value (2.54). 
This way we can conclude that Diet 2 was significantly more effective than Diet 1.


## ANOVA

**ANOVA** (*Analysis of Variance*) is a statistical technique used to compare the measures of two or more groups. The idea behind ANOVA is to decompose the total variability in the data into two components: between-group variability and within-group variability:

- **Between-group variability**: This variability refers to the differences between the group means. If this variability is considerably larger than the within-group variability, it could be an indication that at least one of the group means is different.
- **Within-group variability**: This variability refers to the dispersion of the data within each group. If all groups have similar variability, then any noticeable difference in group means could be considered significant.

Hypotheses in ANOVA typically include:

- **Null hypothesis** ($H₀$): The means of all groups are equal.
- **Alternative hypothesis** ($H₁$): At least one of the group means is different.

If the ANOVA test result is significant (e.g., a p-value less than a threshold such as 0.05), this suggests that at least one group mean is different.

## Exercise 2

A farmer decides to test three different types of fertilizers to determine if one is superior in terms of corn production. The farmer plants corn on 15 identical plots and uses all three fertilizers (5 plots for each type). At the end of the season, he measures the corn yield (in kg) of each plot, with the following result:

| Fertilizer 1 | Fertilizer 2 | Fertilizer 3 |
|:-------------|:-------------|:-------------|
| 20 | 22 | 24 |
| 21 | 21 | 23 |
| 20 | 23 | 22 |
| 19 | 22 | 23 |
| 20 | 21 | 24 |

With this data, he seeks to answer the following question: Is there a significant difference in average corn yield between the three types of fertilizers?

To help you, follow the points below:

- State the hypothesis: null and alternative hypothesis.
- Perform the ANOVA test.
- Analyze the conclusions.
- If one fertilizer is better than another, how can we know it?

In [59]:
# TODO

from scipy.stats import f_oneway as anova

Hypothesis (null):

-The will not be any significant differences between the measures of the corn yields.

Hypothesis 1:

-The will be significant differences between the measures of the corn yields for at least one fertilizer.

In [74]:
fert1 = np.array([20,21,20,19,20])
fert2 = np.array([22,21,23,22,21])
fert3 = np.array([24,23,22,23,24])
data = pd.DataFrame([fert1,fert2,fert3])
data = data.T
data.columns = ("Fertilizer 1", "Fertilizer 2", "Fertilizer 3")
data

Fertilizer 1  Fertilizer 2  Fertilizer 3
0            20            22            24
1            21            21            23
2            20            23            22
3            19            22            23
4            20            21            24

In [69]:
results = anova(data["Fertilizer 1"],data["Fertilizer 2"], data["Fertilizer 3"])
results


F_onewayResult(statistic=np.float64(20.315789473684188), pvalue=np.float64(0.00014047824793190475))

In [71]:
"""
To analyze the differences between the results of both diets a Two Sample Student's T Test was performed.
"""

print("Conclusions:")
print(f"Results from the Anova Test show that the differences in outcomes between the three fertilizers are very significant, with a p value lesser than .05 ({round(results[1],4)}). Therefore, confirming Hyphothesis 1. \n")

Conclusions:
Results from the Anova Test show that the differences in outcomes between the three fertilizers are very significant, with a p value lesser than .05 (0.0001)). Therefore, confirming Hyphothesis 1. 



In [100]:
"""
Furthermore, Two Sample Student's T tests were performed between each of the groups to stablish which of them had statistically significant differences.
"""
#Differences between Fertilizer 1, 2 and 3
tests = pd.DataFrame([[t(a=fert1, b=fert2).pvalue, t(a=fert1, b=fert3).pvalue, t(a=fert2, b=fert3).pvalue]], columns = ["F1 v F2", "F1 v F3", "F2 v F3"], index = ["P Value"])
tests


F1 v F2   F1 v F3   F2 v F3
P Value  0.006271  0.000182  0.029449

In [101]:
"""
Now let's analyze the mean value for each Fertilizer
"""
means = pd.DataFrame([[np.mean(fert1), np.mean(fert2), np.mean(fert3)]], columns = [data.columns], index = ["Means"])
means

Fertilizer 1 Fertilizer 2 Fertilizer 3
Means         20.0         21.8         23.2

Conclusions:

Results from the ANOVA test show that there are statistical differences in the values of at least one of the corn yields with a p value of 0.0001. 
Further analysis showed that there is significant differences between the corn yield of every group with the Fertilizer 3 having the biggest mean value (23.2), and the Fertilizer 1 the smallest (20.0)